File: ios/Navi/Services/MapboxManager.swift
import Foundation
import MapboxMaps
import CoreLocation
import Combine

/// Mapbox Manager - Handles all Mapbox SDK interactions for map rendering and navigation
class MapboxManager: ObservableObject {
    static let shared = MapboxManager()
    
    // MARK: - Published Properties
    @Published var mapView: MapView?
    @Published var currentRoute: Route?
    @Published var isNavigating = false
    @Published var currentInstruction: RouteInstruction?
    @Published var remainingDistance: Double = 0
    @Published var remainingDuration: Double = 0
    @Published var error: MapboxError?
    
    // MARK: - Private Properties
    private var pointAnnotationManager: PointAnnotationManager?
    private var polylineAnnotationManager: PolylineAnnotationManager?
    private var cancellables = Set<AnyCancellable>()
    
    // Mapbox Access Token - Replace with your token
    private let accessToken = "YOUR_MAPBOX_ACCESS_TOKEN"
    
    // MARK: - Initialization
    private init() {
        // Configure Mapbox
        MapboxOptions.accessToken = accessToken
    }
    
    // MARK: - Map Setup
    
    /// Initialize map view with default configuration
    func createMapView(frame: CGRect) -> MapView {
        let resourceOptions = ResourceOptions(accessToken: accessToken)
        let mapInitOptions = MapInitOptions(
            resourceOptions: resourceOptions,
            styleURI: .streets
        )
        
        let mapView = MapView(frame: frame, mapInitOptions: mapInitOptions)
        self.mapView = mapView
        
        // Setup annotation managers
        setupAnnotationManagers()
        
        return mapView
    }
    
    /// Setup annotation managers for markers and routes
    private func setupAnnotationManagers() {
        guard let mapView = mapView else { return }
        
        pointAnnotationManager = mapView.annotations.makePointAnnotationManager()
        polylineAnnotationManager = mapView.annotations.makePolylineAnnotationManager()
    }
    
    // MARK: - Camera Control
    
    /// Center map on user location
    func centerOnUserLocation(_ location: CLLocationCoordinate2D, zoom: Double = 15.0) {
        guard let mapView = mapView else { return }
        
        let cameraOptions = CameraOptions(
            center: location,
            zoom: zoom,
            bearing: 0,
            pitch: 0
        )
        
        mapView.camera.ease(to: cameraOptions, duration: 1.0)
    }
    
    /// Fit map to show entire route
    func fitRouteInView(route: Route, padding: UIEdgeInsets = UIEdgeInsets(top: 100, left: 50, bottom: 100, right: 50)) {
        guard let mapView = mapView else { return }
        
        let coordinates = route.geometry.coordinates
        guard !coordinates.isEmpty else { return }
        
        let coordinateBounds = CoordinateBounds(coordinates: coordinates)
        
        let cameraOptions = mapView.mapboxMap.camera(
            for: coordinateBounds,
            padding: padding,
            bearing: nil,
            pitch: nil
        )
        
        mapView.camera.ease(to: cameraOptions, duration: 1.0)
    }
    
    // MARK: - Markers
    
    /// Add a marker at a location
    func addMarker(at coordinate: CLLocationCoordinate2D, title: String? = nil, color: UIColor = .red) {
        guard let pointAnnotationManager = pointAnnotationManager else { return }
        
        var annotation = PointAnnotation(coordinate: coordinate)
        annotation.textField = title
        annotation.textColor = StyleColor(.black)
        annotation.iconImage = "marker-icon"
        
        pointAnnotationManager.annotations.append(annotation)
    }
    
    /// Clear all markers
    func clearMarkers() {
        pointAnnotationManager?.annotations.removeAll()
    }
    
    // MARK: - Route Display
    
    /// Display route on map
    func displayRoute(_ route: Route) {
        guard let polylineAnnotationManager = polylineAnnotationManager else { return }
        
        // Clear existing routes
        polylineAnnotationManager.annotations.removeAll()
        
        // Create polyline from route geometry
        var polyline = PolylineAnnotation(lineCoordinates: route.geometry.coordinates)
        polyline.lineColor = StyleColor(UIColor(red: 0.15, green: 0.39, blue: 0.92, alpha: 1.0)) // Navi blue
        polyline.lineWidth = 6.0
        polyline.lineJoin = .round
        polyline.lineCap = .round
        
        polylineAnnotationManager.annotations.append(polyline)
        
        // Add start and end markers
        if let start = route.geometry.coordinates.first {
            addMarker(at: start, title: "Start", color: .green)
        }
        if let end = route.geometry.coordinates.last {
            addMarker(at: end, title: "Destination", color: .red)
        }
        
        // Fit route in view
        fitRouteInView(route: route)
        
        self.currentRoute = route
    }
    
    /// Clear route from map
    func clearRoute() {
        polylineAnnotationManager?.annotations.removeAll()
        clearMarkers()
        currentRoute = nil
    }
    
    // MARK: - Route Calculation
    
    /// Calculate route between two points
    func calculateRoute(from origin: CLLocationCoordinate2D, to destination: CLLocationCoordinate2D, completion: @escaping (Result<Route, MapboxError>) -> Void) {
        
        let originString = "\(origin.longitude),\(origin.latitude)"
        let destinationString = "\(destination.longitude),\(destination.latitude)"
        
        let urlString = "https://api.mapbox.com/directions/v5/mapbox/driving/\(originString);\(destinationString)?geometries=geojson&steps=true&access_token=\(accessToken)"
        
        guard let url = URL(string: urlString) else {
            completion(.failure(.invalidURL))
            return
        }
        
        URLSession.shared.dataTask(with: url) { [weak self] data, response, error in
            if let error = error {
                DispatchQueue.main.async {
                    completion(.failure(.networkError(error.localizedDescription)))
                }
                return
            }
            
            guard let data = data else {
                DispatchQueue.main.async {
                    completion(.failure(.noData))
                }
                return
            }
            
            do {
                let decoder = JSONDecoder()
                let directionsResponse = try decoder.decode(DirectionsResponse.self, from: data)
                
                guard let route = directionsResponse.routes.first else {
                    DispatchQueue.main.async {
                        completion(.failure(.noRouteFound))
                    }
                    return
                }
                
                DispatchQueue.main.async {
                    self?.currentRoute = route
                    completion(.success(route))
                }
            } catch {
                DispatchQueue.main.async {
                    completion(.failure(.decodingError(error.localizedDescription)))
                }
            }
        }.resume()
    }
    
    /// Calculate route with waypoints
    func calculateRouteWithWaypoints(waypoints: [CLLocationCoordinate2D], completion: @escaping (Result<Route, MapboxError>) -> Void) {
        
        guard waypoints.count >= 2 else {
            completion(.failure(.invalidWaypoints))
            return
        }
        
        let coordinatesString = waypoints.map { "\($0.longitude),\($0.latitude)" }.joined(separator: ";")
        
        let urlString = "https://api.mapbox.com/directions/v5/mapbox/driving/\(coordinatesString)?geometries=geojson&steps=true&access_token=\(accessToken)"
        
        guard let url = URL(string: urlString) else {
            completion(.failure(.invalidURL))
            return
        }
        
        URLSession.shared.dataTask(with: url) { [weak self] data, response, error in
            if let error = error {
                DispatchQueue.main.async {
                    completion(.failure(.networkError(error.localizedDescription)))
                }
                return
            }
            
            guard let data = data else {
                DispatchQueue.main.async {
                    completion(.failure(.noData))
                }
                return
            }
            
            do {
                let decoder = JSONDecoder()
                let directionsResponse = try decoder.decode(DirectionsResponse.self, from: data)
                
                guard let route = directionsResponse.routes.first else {
                    DispatchQueue.main.async {
                        completion(.failure(.noRouteFound))
                    }
                    return
                }
                
                DispatchQueue.main.async {
                    self?.currentRoute = route
                    completion(.success(route))
                }
            } catch {
                DispatchQueue.main.async {
                    completion(.failure(.decodingError(error.localizedDescription)))
                }
            }
        }.resume()
    }
    
    // MARK: - Map Style
    
    /// Change map style
    func setMapStyle(_ style: MapStyle) {
        guard let mapView = mapView else { return }
        
        let styleURI: StyleURI
        switch style {
        case .streets:
            styleURI = .streets
        case .satellite:
            styleURI = .satellite
        case .satelliteStreets:
            styleURI = .satelliteStreets
        case .dark:
            styleURI = .dark
        case .light:
            styleURI = .light
        case .outdoors:
            styleURI = .outdoors
        }
        
        mapView.mapboxMap.loadStyleURI(styleURI)
    }
    
    // MARK: - Traffic
    
    /// Toggle traffic layer
    func toggleTraffic(enabled: Bool) {
        guard let mapView = mapView else { return }
        
        if enabled {
            // Add traffic layer
            try? mapView.mapboxMap.style.addLayer(TrafficLayer())
        } else {
            // Remove traffic layer
            try? mapView.mapboxMap.style.removeLayer(withId: "traffic")
        }
    }
}

// MARK: - Supporting Types

/// Map style options
enum MapStyle {
    case streets
    case satellite
    case satelliteStreets
    case dark
    case light
    case outdoors
}

/// Mapbox error types
enum MapboxError: Error, LocalizedError {
    case invalidURL
    case networkError(String)
    case noData
    case noRouteFound
    case decodingError(String)
    case invalidWaypoints
    
    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid URL for route calculation"
        case .networkError(let message):
            return "Network error: \(message)"
        case .noData:
            return "No data received from Mapbox"
        case .noRouteFound:
            return "No route found between the locations"
        case .decodingError(let message):
            return "Failed to decode response: \(message)"
        case .invalidWaypoints:
            return "Invalid waypoints - need at least 2 points"
        }
    }
}

// MARK: - Directions API Models

struct DirectionsResponse: Codable {
    let routes: [Route]
    let waypoints: [Waypoint]?
    let code: String
}

struct Route: Codable {
    let distance: Double // meters
    let duration: Double // seconds
    let geometry: Geometry
    let legs: [RouteLeg]
    let weight: Double?
    let weightName: String?
}

struct Geometry: Codable {
    let coordinates: [CLLocationCoordinate2D]
    let type: String
    
    enum CodingKeys: String, CodingKey {
        case coordinates
        case type
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        type = try container.decode(String.self, forKey: .type)
        
        let coordArrays = try container.decode([[Double]].self, forKey: .coordinates)
        coordinates = coordArrays.map { CLLocationCoordinate2D(latitude: $0[1], longitude: $0[0]) }
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(type, forKey: .type)
        let coordArrays = coordinates.map { [$0.longitude, $0.latitude] }
        try container.encode(coordArrays, forKey: .coordinates)
    }
}

struct RouteLeg: Codable {
    let distance: Double
    let duration: Double
    let steps: [RouteStep]
    let summary: String?
}

struct RouteStep: Codable {
    let distance: Double
    let duration: Double
    let geometry: Geometry
    let name: String?
    let mode: String
    let maneuver: Maneuver
}

struct Maneuver: Codable {
    let location: [Double]
    let bearingBefore: Double?
    let bearingAfter: Double?
    let instruction: String
    let type: String
    let modifier: String?
    
    enum CodingKeys: String, CodingKey {
        case location
        case bearingBefore = "bearing_before"
        case bearingAfter = "bearing_after"
        case instruction
        case type
        case modifier
    }
}

struct Waypoint: Codable {
    let name: String
    let location: [Double]
}

// MARK: - Route Instruction (for navigation)
struct RouteInstruction {
    let text: String
    let distance: Double
    let type: String
    let modifier: String?
    let location: CLLocationCoordinate2D
}

// MARK: - Traffic Layer (placeholder)
struct TrafficLayer: Layer {
    var id: String = "traffic"
    var type: LayerType = .line
    var filter: Expression?
    var source: String?
    var sourceLayer: String?
    var minZoom: Double?
    var maxZoom: Double?
}

// MARK: - CoordinateBounds Extension
extension CoordinateBounds {
    init(coordinates: [CLLocationCoordinate2D]) {
        guard !coordinates.isEmpty else {
            self.init(southwest: CLLocationCoordinate2D(latitude: 0, longitude: 0),
                     northeast: CLLocationCoordinate2D(latitude: 0, longitude: 0))
            return
        }
        
        var minLat = coordinates[0].latitude
        var maxLat = coordinates[0].latitude
        var minLon = coordinates[0].longitude
        var maxLon = coordinates[0].longitude
        
        for coord in coordinates {
            minLat = min(minLat, coord.latitude)
            maxLat = max(maxLat, coord.latitude)
            minLon = min(minLon, coord.longitude)
            maxLon = max(maxLon, coord.longitude)
        }
        
        self.init(
            southwest: CLLocationCoordinate2D(latitude: minLat, longitude: minLon),
            northeast: CLLocationCoordinate2D(latitude: maxLat, longitude: maxLon)
        )
    }
}

File: android/app/src/main/java/com/navi/data/services/MapboxManager.kt
package com.navi.data.services

import android.content.Context
import android.graphics.Color
import com.mapbox.geojson.LineString
import com.mapbox.geojson.Point
import com.mapbox.maps.CameraOptions
import com.mapbox.maps.EdgeInsets
import com.mapbox.maps.MapView
import com.mapbox.maps.Style
import com.mapbox.maps.plugin.annotation.annotations
import com.mapbox.maps.plugin.annotation.generated.*
import com.mapbox.maps.plugin.locationcomponent.location
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.OkHttpClient
import okhttp3.Request
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.max
import kotlin.math.min

@Singleton
class MapboxManager @Inject constructor(
    @ApplicationContext private val context: Context
) {
    // Mapbox Access Token - Replace with your token
    private val accessToken = "YOUR_MAPBOX_ACCESS_TOKEN"
    
    // State flows
    private val _currentRoute = MutableStateFlow<Route?>(null)
    val currentRoute: StateFlow<Route?> = _currentRoute.asStateFlow()
    
    private val _isNavigating = MutableStateFlow(false)
    val isNavigating: StateFlow<Boolean> = _isNavigating.asStateFlow()
    
    private val _currentInstruction = MutableStateFlow<RouteInstruction?>(null)
    val currentInstruction: StateFlow<RouteInstruction?> = _currentInstruction.asStateFlow()
    
    private val _remainingDistance = MutableStateFlow(0.0)
    val remainingDistance: StateFlow<Double> = _remainingDistance.asStateFlow()
    
    private val _remainingDuration = MutableStateFlow(0.0)
    val remainingDuration: StateFlow<Double> = _remainingDuration.asStateFlow()
    
    private val _error = MutableStateFlow<MapboxError?>(null)
    val error: StateFlow<MapboxError?> = _error.asStateFlow()
    
    // Annotation managers
    private var pointAnnotationManager: PointAnnotationManager? = null
    private var polylineAnnotationManager: PolylineAnnotationManager? = null
    
    private val httpClient = OkHttpClient()
    private val json = Json { ignoreUnknownKeys = true }
    
    // MARK: - Map Setup
    
    fun setupMap(mapView: MapView) {
        // Load map style
        mapView.getMapboxMap().loadStyleUri(Style.MAPBOX_STREETS)
        
        // Setup annotation managers
        pointAnnotationManager = mapView.annotations.createPointAnnotationManager()
        polylineAnnotationManager = mapView.annotations.createPolylineAnnotationManager()
        
        // Enable location component
        mapView.location.updateSettings {
            enabled = true
            pulsingEnabled = true
        }
    }
    
    // MARK: - Camera Control
    
    fun centerOnLocation(mapView: MapView, latitude: Double, longitude: Double, zoom: Double = 15.0) {
        val cameraOptions = CameraOptions.Builder()
            .center(Point.fromLngLat(longitude, latitude))
            .zoom(zoom)
            .build()
        
        mapView.getMapboxMap().setCamera(cameraOptions)
    }
    
    fun fitRouteInView(mapView: MapView, route: Route, padding: EdgeInsets = EdgeInsets(100.0, 50.0, 100.0, 50.0)) {
        val coordinates = route.geometry.coordinates
        if (coordinates.isEmpty()) return
        
        val bounds = calculateBounds(coordinates)
        
        val cameraOptions = mapView.getMapboxMap().cameraForCoordinateBounds(
            bounds,
            padding
        )
        
        mapView.getMapboxMap().setCamera(cameraOptions)
    }
    
    private fun calculateBounds(coordinates: List<Point>): com.mapbox.geojson.BoundingBox {
        var minLat = coordinates[0].latitude()
        var maxLat = coordinates[0].latitude()
        var minLon = coordinates[0].longitude()
        var maxLon = coordinates[0].longitude()
        
        for (coord in coordinates) {
            minLat = min(minLat, coord.latitude())
            maxLat = max(maxLat, coord.latitude())
            minLon = min(minLon, coord.longitude())
            maxLon = max(maxLon, coord.longitude())
        }
        
        return com.mapbox.geojson.BoundingBox.fromLngLats(minLon, minLat, maxLon, maxLat)
    }
    
    // MARK: - Markers
    
    fun addMarker(latitude: Double, longitude: Double, title: String? = null) {
        val pointAnnotation = PointAnnotationOptions()
            .withPoint(Point.fromLngLat(longitude, latitude))
            .withIconImage("marker-icon")
        
        title?.let {
            pointAnnotation.withTextField(it)
        }
        
        pointAnnotationManager?.create(pointAnnotation)
    }
    
    fun clearMarkers() {
        pointAnnotationManager?.deleteAll()
    }
    
    // MARK: - Route Display
    
    fun displayRoute(mapView: MapView, route: Route) {
        // Clear existing routes
        polylineAnnotationManager?.deleteAll()
        clearMarkers()
        
        // Create polyline from route geometry
        val lineString = LineString.fromLngLats(route.geometry.coordinates)
        
        val polylineAnnotation = PolylineAnnotationOptions()
            .withLineString(lineString)
            .withLineColor(Color.parseColor("#2563EB")) // Navi blue
            .withLineWidth(6.0)
        
        polylineAnnotationManager?.create(polylineAnnotation)
        
        // Add start and end markers
        route.geometry.coordinates.firstOrNull()?.let { start ->
            addMarker(start.latitude(), start.longitude(), "Start")
        }
        route.geometry.coordinates.lastOrNull()?.let { end ->
            addMarker(end.latitude(), end.longitude(), "Destination")
        }
        
        // Fit route in view
        fitRouteInView(mapView, route)
        
        _currentRoute.value = route
    }
    
    fun clearRoute() {
        polylineAnnotationManager?.deleteAll()
        clearMarkers()
        _currentRoute.value = null
    }
    
    // MARK: - Route Calculation
    
    suspend fun calculateRoute(
        originLat: Double,
        originLon: Double,
        destLat: Double,
        destLon: Double
    ): Result<Route> {
        return try {
            val originString = "$originLon,$originLat"
            val destString = "$destLon,$destLat"
            
            val url = "https://api.mapbox.com/directions/v5/mapbox/driving/$originString;$destString?geometries=geojson&steps=true&access_token=$accessToken"
            
            val request = Request.Builder()
                .url(url)
                .build()
            
            val response = httpClient.newCall(request).execute()
            
            if (!response.isSuccessful) {
                return Result.failure(MapboxError.NetworkError("HTTP ${response.code}"))
            }
            
            val responseBody = response.body?.string() ?: return Result.failure(MapboxError.NoData)
            
            val directionsResponse = json.decodeFromString<DirectionsResponse>(responseBody)
            
            val route = directionsResponse.routes.firstOrNull()
                ?: return Result.failure(MapboxError.NoRouteFound)
            
            _currentRoute.value = route
            Result.success(route)
        } catch (e: Exception) {
            Result.failure(MapboxError.DecodingError(e.message ?: "Unknown error"))
        }
    }
    
    suspend fun calculateRouteWithWaypoints(waypoints: List<Pair<Double, Double>>): Result<Route> {
        if (waypoints.size < 2) {
            return Result.failure(MapboxError.InvalidWaypoints)
        }
        
        return try {
            val coordinatesString = waypoints.joinToString(";") { "${it.second},${it.first}" }
            
            val url = "https://api.mapbox.com/directions/v5/mapbox/driving/$coordinatesString?geometries=geojson&steps=true&access_token=$accessToken"
            
            val request = Request.Builder()
                .url(url)
                .build()
            
            val response = httpClient.newCall(request).execute()
            
            if (!response.isSuccessful) {
                return Result.failure(MapboxError.NetworkError("HTTP ${response.code}"))
            }
            
            val responseBody = response.body?.string() ?: return Result.failure(MapboxError.NoData)
            
            val directionsResponse = json.decodeFromString<DirectionsResponse>(responseBody)
            
            val route = directionsResponse.routes.firstOrNull()
                ?: return Result.failure(MapboxError.NoRouteFound)
            
            _currentRoute.value = route
            Result.success(route)
        } catch (e: Exception) {
            Result.failure(MapboxError.DecodingError(e.message ?: "Unknown error"))
        }
    }
    
    // MARK: - Map Style
    
    fun setMapStyle(mapView: MapView, style: MapStyle) {
        val styleUri = when (style) {
            MapStyle.STREETS -> Style.MAPBOX_STREETS
            MapStyle.SATELLITE -> Style.SATELLITE
            MapStyle.SATELLITE_STREETS -> Style.SATELLITE_STREETS
            MapStyle.DARK -> Style.DARK
            MapStyle.LIGHT -> Style.LIGHT
            MapStyle.OUTDOORS -> Style.OUTDOORS
        }
        
        mapView.getMapboxMap().loadStyleUri(styleUri)
    }
}

// MARK: - Supporting Types

enum class MapStyle {
    STREETS,
    SATELLITE,
    SATELLITE_STREETS,
    DARK,
    LIGHT,
    OUTDOORS
}

sealed class MapboxError : Exception() {
    object InvalidURL : MapboxError()
    data class NetworkError(val msg: String) : MapboxError()
    object NoData : MapboxError()
    object NoRouteFound : MapboxError()
    data class DecodingError(val msg: String) : MapboxError()
    object InvalidWaypoints : MapboxError()
    
    fun getMessage(): String = when (this) {
        is InvalidURL -> "Invalid URL for route calculation"
        is NetworkError -> "Network error: $msg"
        is NoData -> "No data received from Mapbox"
        is NoRouteFound -> "No route found between the locations"
        is DecodingError -> "Failed to decode response: $msg"
        is InvalidWaypoints -> "Invalid waypoints - need at least 2 points"
    }
}

// MARK: - Directions API Models

@Serializable
data class DirectionsResponse(
    val routes: List<Route>,
    val waypoints: List<Waypoint>? = null,
    val code: String
)

@Serializable
data class Route(
    val distance: Double, // meters
    val duration: Double, // seconds
    val geometry: Geometry,
    val legs: List<RouteLeg>,
    val weight: Double? = null,
    val weightName: String? = null
)

@Serializable
data class Geometry(
    val coordinates: List<@Serializable(with = PointSerializer::class) Point>,
    val type: String
)

@Serializable
data class RouteLeg(
    val distance: Double,
    val duration: Double,
    val steps: List<RouteStep>,
    val summary: String? = null
)

@Serializable
data class RouteStep(
    val distance: Double,
    val duration: Double,
    val geometry: Geometry,
    val name: String? = null,
    val mode: String,
    val maneuver: Maneuver
)

@Serializable
data class Maneuver(
    val location: List<Double>,
    val bearingBefore: Double? = null,
    val bearingAfter: Double? = null,
    val instruction: String,
    val type: String,
    val modifier: String? = null
)

@Serializable
data class Waypoint(
    val name: String,
    val location: List<Double>
)

// Route Instruction for navigation
data class RouteInstruction(
    val text: String,
    val distance: Double,
    val type: String,
    val modifier: String?,
    val latitude: Double,
    val longitude: Double
)

// Point Serializer for Mapbox Point
object PointSerializer : kotlinx.serialization.KSerializer<Point> {
    override val descriptor = kotlinx.serialization.descriptors.PrimitiveSerialDescriptor("Point", kotlinx.serialization.descriptors.PrimitiveKind.STRING)
    
    override fun serialize(encoder: kotlinx.serialization.encoding.Encoder, value: Point) {
        encoder.encodeString("${value.longitude()},${value.latitude()}")
    }
    
    override fun deserialize(decoder: kotlinx.serialization.encoding.Decoder): Point {
        val coords = decoder.decodeString().split(",")
        return Point.fromLngLat(coords[0].toDouble(), coords[1].toDouble())
    }
}
